sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -90,6 +90,7 @@ trait PredicateHelper {
    * Returns true iff `expr` could be evaluated as a condition within join.
    */
   protected def canEvaluateWithinJoin(expr: Expression): Boolean = expr match {
+    case l: ListQuery => false
```

Comment:
@viirya Now pulling up IN predicates is deferred to `RewritePredicateSubquery`. As a result, `ListQuery`'s children can be empty before that rule and falls into `case e: SubqueryExpression`.